packages/playground/blueprints/src/lib/zip.ts

```diff
@@ -0,0 +1,170 @@
+	startsAt?: number;
+	extract?: any;
+	signature: string;
+	version: number;
+	generalPurpose: number;
+	compressionMethod: number;
+	lastModifiedTime: number;
+	lastModifiedDate: number;
+	crc: number;
+	compressedSize: number;
+	uncompressedSize: number;
+	fileNameLength: number;
+	fileName: string;
+	extraLength: number;
+	extra: string;
+}
+
+export interface ZipFileEntry extends ZipFileHeader {
+	dataStream: ReadableStream<Uint8Array>;
+}
+
+const SIGNATURE_FILE = 0x04034b50;
+const SIGNATURE_CENTRAL_DIRECTORY_START = 0x02014b50;
+const SIGNATURE_CENTRAL_DIRECTORY_END = 0x06054b50;
+
+async function* iterateZipEntries(stream: ReadableStream<Uint8Array>) {
+	const zipStream = new BufferedReadableStream(stream.getReader());
+	while (true) {
+		const signature = await zipStream.readUint32();
+		// console.log({ signature });
+		if (signature === SIGNATURE_FILE) {
+			const header = await readFileHeader(zipStream);
+			const fileDataStream = await createFileDataStream(
+				zipStream,
+				header
+			);
+			yield {
+				...header,
+				dataStream: fileDataStream,
+			} as ZipFileEntry;
+		} else if (signature === SIGNATURE_CENTRAL_DIRECTORY_START) {
+			break;
+		} else if (signature === SIGNATURE_CENTRAL_DIRECTORY_END) {
+			break;
+		} else {
+			throw new Error(`Unknown signature: ${signature}`);
+		}
+	}
+}
+
+export async function* iterateZipFiles(stream: ReadableStream<Uint8Array>) {
+	for await (const entry of iterateZipEntries(stream)) {
+		yield toZipFile(entry);
+	}
+}
+
+function toZipFile(entry: ZipFileEntry): FileEntry {
+	return {
+		path: entry.fileName,
+		isDirectory: entry.fileName.endsWith('/'),
+		async read() {
+			return await readAllBytes(entry.dataStream.getReader());
+		},
+	};
+}
+
+async function readFileHeader(
+	stream: BufferedReadableStream
+): Promise<ZipFileHeader> {
+	const entry: Partial<ZipFileEntry> = {};
+	entry['version'] = await stream.readUint16();
+	entry['generalPurpose'] = await stream.readUint16();
+	entry['compressionMethod'] = await stream.readUint16();
+	entry['lastModifiedTime'] = await stream.readUint16();
+	entry['lastModifiedDate'] = await stream.readUint16();
+	entry['crc'] = await stream.readUint32();
+	entry['compressedSize'] = await stream.readUint32();
+	entry['uncompressedSize'] = await stream.readUint32();
+	entry['fileNameLength'] = await stream.readUint16();
+	entry['extraLength'] = await stream.readUint16();
+	entry['fileName'] = new TextDecoder().decode(
+		await stream.read(entry.fileNameLength)
+	);
+	entry['extra'] = new TextDecoder().decode(
+		await stream.read(entry['extraLength'])
+	);
+	return entry as ZipFileHeader;
+}
+
+async function createFileDataStream(
+	stream: BufferedReadableStream,
+	header: ZipFileHeader
+): Promise<ReadableStream<Uint8Array>> {
+	if (header.compressedSize === 0) {
+		return new ReadableStream({
+			start(controller) {
+				controller.close();
+			},
+		});
+	}
+
+	const data = await stream.read(header.compressedSize);
```

Comment:
curious: what happens if we attempt to read more bytes than are available?

Reply:
I removed the custom `Stream` class implementation. Right now we read `n` bytes using the native [ReadableStreamBYOBReader](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamBYOBReader) class which sends a "read request" to the data source. Every data source is free to process it how it wants, but typically it will the buffer up `Math.min(Buffer.size, numberOfAvailableDataBytes)`.